stor/s3.py

```diff
@@ -580,25 +580,18 @@ def download(self, dest, condition=None, use_manifest=False, **kwargs):
 
         downloaded = {'completed': [], 'failed': []}
         with S3DownloadLogger(len(files_to_download)) as dl:
-            pool = ThreadPool(options['object_threads'])
-            try:
-                result_iter = pool.imap_unordered(download_w_config, files_to_download)
-                while True:
-                    try:
-                        result = result_iter.next(0xFFFF)
-                        if result['success']:
-                            dl.add_result(result)
-                            downloaded['completed'].append(result)
-                        else:
-                            downloaded['failed'].append(result)
-                    except StopIteration:
-                        break
-                pool.close()
-            except BaseException:
-                pool.terminate()
-                raise
-            finally:
-                pool.join()
+            with ThreadPoolExecutor(max_workers=options.get("object_threads")) as executor:
+                futures = {
+                    executor.submit(download_w_config, file_to_download): file_to_download
+                    for file_to_download in files_to_download
+                }
+                for fut in as_completed(futures):
+                    result = fut.result()
```

Comment:
any way to catch an exception here and return the file being downloaded which is problematic in the stack trace?
```suggestion
try:
   result = fut.result()
except Exception as e:
    file = future.something(). # not sure how to get file_to_download
    raise SomeGoodErrorMEssageWithFile
```

Reply:
Completed for download in 04cb7b8
Then for upload (with some download test updates) in bcb680d